Use intern for testing2

JacquesPerrault opened the pull request:
pushed commit df72350 to address the first round of feedback

csnover commented on test/dispatcher.js:
```diff
@@ -1,7 +1,19 @@
+define([ 'intern!tdd',
+        'intern/chai!',
```
The dependency lists do not match.

JacquesPerrault replied:
hmmmm...my local looks like this:
define([ 'intern!tdd',
        'intern/chai!',
        'intern/chai!expect',
        '../dist/PEP'
    ],

'git status' says there is nothing to commit, perhaps my rebase pick/squash squashed the wrong thing, but strange that it is still local.

scottgonzalez replied:
That's exactly what's listed here. But you're parameters in the callback don't match the list of dependencies.

JacquesPerrault replied:
fixed.

scottgonzalez commented on test/dispatcher.js:
```diff
@@ -9,11 +21,10 @@ suite('Event Generation and Dispatching', function() {
     document.body.removeChild(container);
   });
 
-  var pepde = PointerEventsPolyfill.dispatcher.eventSources;
```
You can't just reference the module directly here. See above in the capture tests where you have this done correctly.

JacquesPerrault replied:
thanks, I added the line back, and modified it to var pepde = pep.dispatcher.eventSources;

scottgonzalez commented on test/intern.js:
```diff
@@ -0,0 +1,47 @@
+        { browserName: 'firefox', version: '28', platform: [ 'OS X 10.9', 'Windows 7', 'Linux' ] },
+        { browserName: 'chrome', version: '34', platform: [ 'OS X 10.9', 'Windows 7', 'Linux' ] },
+        { browserName: 'safari', version: '6', platform: 'OS X 10.8' },
+        { browserName: 'safari', version: '7', platform: 'OS X 10.9' }
```
@csnover Is there a `version: "latest"` type of API or do we just need to manually bump versions for Chrome and Firefox every once in a while?

csnover replied:
When you’re using a cloud provider, it’s up to them. The last time I checked I think Sauce randomly had some browsers that had a “latest beta” option. If you self-host, then you can just leave the version number off the environment and only ever supply the latest version of browser(s) to the grid. [Capabilities info](http://theintern.github.io/intern/#option-capabilities).

scottgonzalez commented on test/intern.js:
```diff
@@ -0,0 +1,47 @@
+    //suites: [ 'test/loader', 'test/constructor', 'test/dispatcher', 'test/pointermap', 'test/capture' ],
+    suites: [ 'test/setup', 'test/loader', 'test/constructor', 'test/dispatcher', 'test/pointermap', 'test/capture' ],
+    // Functional test suite(s) to run in each browser once non-functional tests are completed
+    functionalSuites: [ 'test/functional/index' ],
```
This file doesn't exist.

JacquesPerrault replied:
Good catch.  I've commented out functionalSuites

scottgonzalez commented on test/intern.js:
```diff
@@ -0,0 +1,47 @@
+    // Functional test suite(s) to run in each browser once non-functional tests are completed
+    functionalSuites: [ 'test/functional/index' ],
+    // A regular expression matching URLs to files that should not be included in code coverage analysis
+    excludeInstrumentation: /^(?:test|node_modules)\//
```
This is probably only supposed to include test, since we're not loading anything form node_modules. I'd need to see the output though to know what's currently being included.

JacquesPerrault replied:
This was an artifact of the default configuration; removing it doesn't seem to harm anything,

scottgonzalez commented on test/loader.js:
```diff
@@ -1,11 +1,19 @@
-    var pep = window.PointerEventsPolyfill;
     test('PointerEventsPolyfill', function() {
-      expect(pep).to.be.ok;
+        expect(pep).to.be.ok;
```
bad indentation

JacquesPerrault replied:
fixed

scottgonzalez commented on test/setup.js:
```diff
@@ -1,5 +1,5 @@
-chai.config.includeStack = true;
-expect = chai.expect;
+//chai.config.includeStack = true;
+//expect = chai.expect;
```
Delete these.

JacquesPerrault replied:
fixed.

scottgonzalez commented on test/dispatcher.js:
```diff
@@ -62,7 +75,7 @@ suite('Event Generation and Dispatching', function() {
   test('PointerEvents from mouse fire anywhere by default', function() {
     // move always fires
-    var cb = chai.spy();
+    var cb = chai.spy;
```
What changed about `chai.spy` that this seems to have been a factory before, but no longer is?

JacquesPerrault replied:
As a factory, this line throws "TypeError: chai.spy is not a function", perhaps that is symptomatic of another problem?

scottgonzalez replied:
I don't see where you're loading the spy functionality. The existing tests have to load that separately: https://github.com/jquery/PEP/blob/67d1dcad49774522ea8bd852b03d6d535883567c/test/runner.html#L10.

scottgonzalez commented:
Tests no longer receive a done callback, so tests like https://github.com/jquery/PEP/blob/67d1dcad49774522ea8bd852b03d6d535883567c/test/capture.js#L36 are now throwing errors when they try to call the `done` callback.

@snover Do you know why this happens? Also Intern detects the errors and logs them to the console, but doesn't seem to care about letting you know in the UI. Is there some flag that needs to be enabled to make Intern mark the test suite as failing if there are errors?

csnover commented:
Yes.

To that specific test you are referencing, an error calling undefined function `done` would be caught and handled by Intern if it was following the first code branch (`navigator.msPointerEnabled`).

However, in the `else` branch, as you note, it does not. This is because Intern was deliberatly designed to not associate uncaught exceptions with specific tests by default, in the interest of leaving the system free to add test parallelisation in the future (without async stack information from the JS engine, an error can’t be automatically reassociated with the original test if an unhandled exception is thrown when multiple tests are running at the same time). This is an area where improvements could definitely be made, but it’s not actually necessary to change Intern to make this work.

To handle the async throws from PEP, simply wrap the async test functions with a wrapper that sets up the right global error handler. One very easy way to do this would be to change the test interface dependency:

```js
define([ './support/interface' ], function (tdd) {
  // ... the tests ...
});
```

and then in support/interface.js:

```js
define([ 'intern!tdd' ], function (tdd) {
  function catchGlobalErrors(testFn) {
    return function () {
      var promise = testFn.apply(this, arguments);

      // testFn.length is checked ONLY to ensure we wrap functions that
      // were written to receive a `done` function so they can be fixed
      if (promise && promise.then || this.isAsync || testFn.length > 0) {
        var dfd = this.async();

        if (promise && promise !== dfd.promise) {
          promise.then(dfd.resolve.bind(dfd), dfd.reject.bind(dfd));
        }

        var originalHandler = window.onerror;
        window.onerror = function (message, url, line, column, error) {
          dfd.reject(error || new Error(message));
        };

        return dfd.promise.then(function (returnValue) {
          window.onerror = originalHandler;
          return returnValue;
        }, function (error) {
          window.onerror = originalHandler;
          throw error;
        });
      }
    };
  }

  var autoCapturingTdd = Object.create(tdd);
  autoCapturingTdd.test = function (name, test) {
    tdd.test(name, catchGlobalErrors(test));
  };

  return autoCapturingTdd;
});
```

It would probably be useful to make this an opt-in convenience within Intern itself, so you may open another ticket about that if you want. (One may actually exist already, I’m not sure.)

Now, why does it not show up in the final report? In the case that an unhandled exception occurs, Intern publishes an error topic that needs to be handled separately from normal test failures. Apparently, the html reporter is not set up to listen for those. Please open a bug on this, since it should not say everything’s OK when it’s not OK.

scottgonzalez commented:
@JacquesPerrault Can you add that support interface (without the ES5 dependency, so no `Function#bind()` usage?

We'll also need to either make this provide a done callback where needed or update the tests to return promises.

JacquesPerrault commented:
@snover - thanks for the explanation - I've been looking at that for a couple days and couldn't figure out what was going on. 
@scottgonzalez - I'll get to work on the interface

csnover commented:
Please feel free to reach out to me on IRC or on GitHub if you get stuck again.

scottgonzalez commented on test/intern.js:
```diff
@@ -0,0 +1,46 @@
+        packages: [ { name: 'app', location: '.' } ]
+    },
+    // Non-functional test suite(s) to run in each browser
+    suites: [ 'test/unit/loader', 'test/unit/constructor', 'test/unit/dispatcher', 'test/unit/pointermap', 'test/unit/capture' ],
```
Let's drop everything above this line; none of it is needed until we have a CI setup and we'll fill in the correct values at that point.

scottgonzalez commented on test/intern.js:
```diff
@@ -0,0 +1,46 @@
+    // Functional test suite(s) to run in each browser once non-functional tests are completed
+    //functionalSuites: [ 'test/functional/index' ],
+    // A regular expression matching URLs to files that should not be included in code coverage analysis
+    excludeInstrumentation: /^(?:test)\//
```
Let's add node_modules back. When this is actually generating coverage, we do need to exclude it. Just tested by running this on BrowserStack.

scottgonzalez replied:
Drop the comments though.

JacquesPerrault commented:
I have added the custom intern test interface and updated capture.js.  However, it is still throwing "callback is not a function" and "done is not a function"  stemming from here:

https://github.com/JacquesPerrault/PEP/blob/333e7c92b913e7f1ca91fec6e4594f8b2e7bb00d/test/unit/capture.js#L121

scottgonzalez commented:
Closing in favor of #167.
